File: src/main/kotlin/com/github/hirsivaja/ip/IpUtils.kt
package com.github.hirsivaja.ip

import java.nio.ByteBuffer
import java.util.logging.Logger

object IpUtils {
    private val LOG: Logger = Logger.getLogger(IpUtils::class.java.name)

    fun calculateInternetChecksum(data: ByteArray): Short {
        val buf: ByteBuffer = ByteBuffer.wrap(data)
        var sum: Long = 0
        while (buf.hasRemaining()) {
            sum += if (buf.remaining() > 1) {
                buf.getShort().toInt() and 0xFFFF
            } else {
                (buf.get().toInt() shl 8) and 0xFFFF
            }
        }
        return ((((sum and 0xFFFF) + (sum shr 16)).inv()) and 0xFFFF).toShort()
    }

    fun verifyInternetChecksum(checksumData: ByteArray): Boolean {
        return verifyInternetChecksum(checksumData, 0.toShort())
    }

    fun verifyInternetChecksum(checksumData: ByteArray, actual: Short): Boolean {
        val expected = calculateInternetChecksum(checksumData)
        if(expected != actual) {
            LOG.warning { "CRC mismatch!" }
        }
        return expected == actual
    }

    fun ensureInternetChecksum(checksumData: ByteArray) {
        ensureInternetChecksum(checksumData, 0.toShort())
    }

    fun ensureInternetChecksum(checksumData: ByteArray, actual: Short) {
        val expected = calculateInternetChecksum(checksumData)
        require(expected == actual) { "The calculated checksum $actual was not the provided checksum $expected" }
    }
}

File: src/main/scala/com/github/hirsivaja/ip/IpUtils.scala
package com.github.hirsivaja.ip

import java.nio.ByteBuffer
import java.util.logging.{Level, Logger}

object IpUtils {
  private val logger = Logger.getLogger("IpUtils")

  def calculateInternetChecksum(data: Array[Byte]): Short = {
    val buf = ByteBuffer.wrap(data)
    var sum = 0
    while (buf.hasRemaining) {
      if (buf.remaining > 1) {
        sum += buf.getShort & 0xFFFF
      } else {
        sum += buf.get << 8 & 0xFFFF
      }
    }
    ((~((sum & 0xFFFF) + (sum >> 16))) & 0xFFFF).toShort
  }

  def verifyInternetChecksum(checksumData: Array[Byte]): Boolean = verifyInternetChecksum(checksumData, 0.toShort)

  def verifyInternetChecksum(checksumData: Array[Byte], actual: Short): Boolean = {
    val expected = calculateInternetChecksum(checksumData)
    if (expected != actual) {
      logger.warning("CRC mismatch!")
    }
    expected == actual
  }

  def ensureInternetChecksum(checksumData: Array[Byte]): Unit = {
    ensureInternetChecksum(checksumData, 0.toShort)
  }

  def ensureInternetChecksum(checksumData: Array[Byte], actual: Short): Unit = {
    val expected = calculateInternetChecksum(checksumData)
    if (expected != actual) {
      logger.log(Level.FINEST, s"Checksum mismatch! Expected checksum $expected. Actual checksum $actual")
      throw new IllegalArgumentException("Checksum does not match!")
    }
  }
}
